### PR TITLE
Respect non 200 status to page static generation response

### LINKED ISSUE #51021: next/navigation notFound() function do not apply HTTP 404 status code
### Verify canary release

- [X] I verified that the issue exists in the latest Next.js canary release

### Provide environment information

```bash
Operating System:
      Platform: darwin
      Arch: x64
      Version: Darwin Kernel Version 22.3.0: Mon Jan 30 20:42:11 PST 2023; root:xnu-8792.81.3~2/RELEASE_X86_64
    Binaries:
      Node: 18.13.0
      npm: 8.19.3
      Yarn: N/A
      pnpm: N/A
    Relevant packages:
      next: 13.4.4
      eslint-config-next: 13.3.0
      react: 18.2.0
      react-dom: 18.2.0
      typescript: 5.0.4
```

### Which area(s) of Next.js are affected? (leave empty if unsure)

App directory (appDir: true), Routing (next/router, next/navigation, next/link)

### Link to the code that reproduces this issue or a replay of the bug

https://codesandbox.io/p/sandbox/cranky-pine-f9s292?file=%2Fapp%2Fpage.tsx%3A1%2C1

### To Reproduce

``` typescript
const BlogPage = async ({ params }: { params: { categorySlug: string; blogSlug: string; locale: string } }) => {
  const blog = (
    await dataService.collection<{ blogs: Blog[] }>('Blog', params.locale, {
      filters: {
        blog_category: {
          slug: { eq: params.categorySlug },
        },
        slug: { eq: params.blogSlug },
      },
    })
  ).blogs?.[0];

  if (!blog) {
    // Here the issue
    notFound();
  }

  return <>{JSON.stringify(blog, null, 2)}</>;
}
```

### Describe the Bug

When notFound is called, the 404 page is correctly displayed, but the status is 200 : 

image

### Expected Behavior

The HTTP status code should be 404

### Which browser are you using? (if relevant)

Chrome Version 114.0.5735.106 (Build officiel) (x86_64)

### How are you deploying your application? (if relevant)

next start

### LINKED ISSUE #62228: 404 page isn't server rendered when using notFound()
### Link to the code that reproduces this issue

https://github.com/nikcio/nextjs-notfound-example

### To Reproduce

1. Create a new project using `npx create-next-app@latest`
2. Create a page that uses the `notFound()` function.
3. Go to that page and see the response in dev tools - Here the Body is only scripts (Client rendered)
4. Go to a different page that doesn't exist and see the response in dev tools - Here the body will be rendered (Server rendered)

Example for the `page.tsx`: 
```tsx
import { notFound } from "next/navigation";

export default function Page() {
  notFound();
}
```

## Expected

I expected that the 404 page was server rendered both when using `next dev` and `next start`. But also most importantly server rendered when triggered using the `notFound()` function.

## Current behavior

Examples from: https://github.com/nikcio/nextjs-notfound-example

image

> `localhost:3000/my-page`
image

> `localhost:3000/dynamic/test`
image

> `localhost:3000/dynamic/my-page`
image

All pages returns a `404` status code

### Production build (`next build && next start`)
image

> `localhost:3000/test`
image

> `localhost:3000/my-page`
image

> `localhost:3000/dynamic/test`
image

> `localhost:3000/dynamic/my-page`
image

> The static page `localhost:3000/my-page` returns status code 200
image

### The rendered result for all requests are the default not found page:
image

### Provide environment information

```bash
Operating System:
  Platform: win32
  Arch: x64
  Version: Windows 11 Enterprise
Binaries:
  Node: 20.9.0
  npm: N/A
  Yarn: N/A
  pnpm: 8.6.10
Relevant Packages:
  next: 14.1.0
  eslint-config-next: 14.1.0
  react: 18.2.0
  react-dom: 18.2.0
  typescript: 5.3.3
Next.js Config:
  output: N/A
```

### Which area(s) are affected? (Select all that apply)

App Router, Routing (next/router, next/navigation, next/link)

### Which stage(s) are affected? (Select all that apply)

next dev (local), next start (local)

### Additional context

We encountered this as an issue when using [next-runtime-env](https://github.com/expatfile/next-runtime-env) that adds a `<script>` tag to the `<head>` of the page. This is a problem because when the whole page is client rendered on the not found page the script tag is added after the page is loaded and the script will never be executed without an explicit call from another component. 

You can replicate this using the `Script` component from Next and using `strategy="beforeInteractive"`.

I've made an example here: https://github.com/nikcio/nextjs-notfound-example/tree/example/script-in-head

image

> Captured in development where no 404 page is server rendered but it's the same behavior when using `next start`

### PR DESCRIPTION
**Pull request by huozhi**
### What

In static generation phase of app page, if there's any case that we're receiving 3xx/4xx status code from the response, we 're setting it into the static generation meta now to make sure they're still returning the same status after build.

### Why

During static generation if there's any 3xx/4xx status code that is set in the response, we should respect to it, such as the ones caused by using `notFound()` to mark as 404 response or `redirect` to mark as `307` response.

Closes NEXT-2895
Fixes #51021
Fixes #62228

### CODE REVIEW

**Comment by ijjk** [comment]
## Failing test suites 
Commit: 7ff54efdd2ad8f34db5d10efc4774ffe08ff9fe6

`TURBOPACK=1 pnpm test-start test/e2e/app-dir/static-generation-status/index.test.ts` (turbopack)
- app-dir static-generation-status > should render the non existed route redirect with status 404

<details>
<summary>Expand output</summary>

  ● app-dir static-generation-status › should render the non existed route redirect with status 404

    expect(received).toBe(expected) // Object.is equality

    Expected: 404
    Received: 200

      18 |
      19 |     it('should render the non existed route redirect with status 404', async () => {
    > 20 |       expect((await next.fetch('/_not-found')).status).toBe(404)
         |                                                        ^
      21 |       expect((await next.fetch('/does-not-exist')).status).toBe(404)
      22 |     })
      23 |   }

      at Object.toBe (e2e/app-dir/static-generation-status/index.test.ts:20:56)

</details>


Read more about building and testing Next.js in [contributing.md](https://github.com/vercel/next.js/blob/canary/contributing.md).

**Comment by ijjk** [comment]
# Stats from current PR

<details open>
<summary><strong>Default Build</strong></summary>

<br/>

<details>
<summary><strong>General</strong> Overall increase ⚠️</summary>

|  | vercel/next.js canary  | vercel/next.js fix/static-generation-status | Change |
| - | - | - | - |
| buildDuration | 14s | 14.1s | N/A |
| buildDurationCached | 7.6s | 6.3s | N/A |
| nodeModulesSize | 198 MB | 198 MB | ⚠️ +3.06 kB |
| nextStartRea..uration (ms) | 398ms | 410ms | N/A |

</details>

<details>
<summary><strong>Client Bundles (main, webpack)</strong></summary>

|  | vercel/next.js canary  | vercel/next.js fix/static-generation-status | Change |
| - | - | - | - |
| 2453-HASH.js gzip | 30.8 kB | 30.8 kB | N/A |
| 3304.HASH.js gzip | 181 B | 181 B | ✓ |
| 3f784ff6-HASH.js gzip | 53.7 kB | 53.7 kB | ✓ |
| 8299-HASH.js gzip | 5.04 kB | 5.04 kB | N/A |
| framework-HASH.js gzip | 45.2 kB | 45.2 kB | ✓ |
| main-app-HASH.js gzip | 242 B | 242 B | ✓ |
| main-HASH.js gzip | 32.2 kB | 32.2 kB | N/A |
| webpack-HASH.js gzip | 1.68 kB | 1.68 kB | N/A |
| Overall change | 99.3 kB | 99.3 kB | ✓ |

</details>

<details>
<summary><strong>Legacy Client Bundles (polyfills)</strong></summary>

|  | vercel/next.js canary  | vercel/next.js fix/static-generation-status | Change |
| - | - | - | - |
| polyfills-HASH.js gzip | 31 kB | 31 kB | ✓ |
| Overall change | 31 kB | 31 kB | ✓ |

</details>

<details>
<summary><strong>Client Pages</strong></summary>

|  | vercel/next.js canary  | vercel/next.js fix/static-generation-status | Change |
| - | - | - | - |
| _app-HASH.js gzip | 196 B | 197 B | N/A |
| _error-HASH.js gzip | 184 B | 184 B | ✓ |
| amp-HASH.js gzip | 505 B | 505 B | ✓ |
| css-HASH.js gzip | 324 B | 325 B | N/A |
| dynamic-HASH.js gzip | 2.5 kB | 2.5 kB | N/A |
| edge-ssr-HASH.js gzip | 258 B | 258 B | ✓ |
| head-HASH.js gzip | 352 B | 352 B | ✓ |
| hooks-HASH.js gzip | 370 B | 371 B | N/A |
| image-HASH.js gzip | 4.21 kB | 4.21 kB | ✓ |
| index-HASH.js gzip | 259 B | 259 B | ✓ |
| link-HASH.js gzip | 2.67 kB | 2.67 kB | N/A |
| routerDirect..HASH.js gzip | 314 B | 312 B | N/A |
| script-HASH.js gzip | 386 B | 386 B | ✓ |
| withRouter-HASH.js gzip | 309 B | 309 B | ✓ |
| 1afbb74e6ecf..834.css gzip | 106 B | 106 B | ✓ |
| Overall change | 6.57 kB | 6.57 kB | ✓ |

</details>

<details>
<summary><strong>Client Build Manifests</strong></summary>

|  | vercel/next.js canary  | vercel/next.js fix/static-generation-status | Change |
| - | - | - | - |
| _buildManifest.js gzip | 481 B | 484 B | N/A |
| Overall change | 0 B | 0 B | ✓ |

</details>

<details>
<summary><strong>Rendered Page Sizes</strong></summary>

|  | vercel/next.js canary  | vercel/next.js fix/static-generation-status | Change |
| - | - | - | - |
| index.html gzip | 529 B | 529 B | ✓ |
| link.html gzip | 541 B | 541 B | ✓ |
| withRouter.html gzip | 524 B | 523 B | N/A |
| Overall change | 1.07 kB | 1.07 kB | ✓ |

</details>

<details>
<summary><strong>Edge SSR bundle Size</strong></summary>

|  | vercel/next.js canary  | vercel/next.js fix/static-generation-status | Change |
| - | - | - | - |
| edge-ssr.js gzip | 95.3 kB | 95.3 kB | N/A |
| page.js gzip | 3.04 kB | 3.04 kB | N/A |
| Overall change | 0 B | 0 B | ✓ |

</details>

<details>
<summary><strong>Middleware size</strong></summary>

|  | vercel/next.js canary  | vercel/next.js fix/static-generation-status | Change |
| - | - | - | - |
| middleware-b..fest.js gzip | 625 B | 628 B | N/A |
| middleware-r..fest.js gzip | 151 B | 151 B | ✓ |
| middleware.js gzip | 25.5 kB | 25.5 kB | N/A |
| edge-runtime..pack.js gzip | 839 B | 839 B | ✓ |
| Overall change | 990 B | 990 B | ✓ |

</details>

<details>
<summary><strong>Next Runtimes</strong></summary>

|  | vercel/next.js canary  | vercel/next.js fix/static-generation-status | Change |
| - | - | - | - |
| app-page-exp...dev.js gzip | 170 kB | 170 kB | ✓ |
| app-page-exp..prod.js gzip | 96.8 kB | 96.8 kB | ✓ |
| app-page-tur..prod.js gzip | 98.5 kB | 98.5 kB | ✓ |
| app-page-tur..prod.js gzip | 92.8 kB | 92.8 kB | ✓ |
| app-page.run...dev.js gzip | 144 kB | 144 kB | ✓ |
| app-page.run..prod.js gzip | 91.3 kB | 91.3 kB | ✓ |
| app-route-ex...dev.js gzip | 21.3 kB | 21.3 kB | ✓ |
| app-route-ex..prod.js gzip | 15 kB | 15 kB | ✓ |
| app-route-tu..prod.js gzip | 15 kB | 15 kB | ✓ |
| app-route-tu..prod.js gzip | 14.8 kB | 14.8 kB | ✓ |
| app-route.ru...dev.js gzip | 21 kB | 21 kB | ✓ |
| app-route.ru..prod.js gzip | 14.8 kB | 14.8 kB | ✓ |
| pages-api-tu..prod.js gzip | 9.55 kB | 9.55 kB | ✓ |
| pages-api.ru...dev.js gzip | 9.82 kB | 9.82 kB | ✓ |
| pages-api.ru..prod.js gzip | 9.55 kB | 9.55 kB | ✓ |
| pages-turbo...prod.js gzip | 22.5 kB | 22.5 kB | ✓ |
| pages.runtim...dev.js gzip | 23.1 kB | 23.1 kB | ✓ |
| pages.runtim..prod.js gzip | 22.4 kB | 22.4 kB | ✓ |
| server.runti..prod.js gzip | 50.9 kB | 50.9 kB | ✓ |
| Overall change | 943 kB | 943 kB | ✓ |

</details>

<details>
<summary><strong>build cache</strong></summary>

|  | vercel/next.js canary  | vercel/next.js fix/static-generation-status | Change |
| - | - | - | - |
| 0.pack gzip | 1.57 MB | 1.57 MB | N/A |
| index.pack gzip | 106 kB | 105 kB | N/A |
| Overall change | 0 B | 0 B | ✓ |

</details>

<details>
<summary><strong>Diff details</strong></summary>

<details>
<summary>Diff for <strong>middleware.js</strong></summary>

Diff too large to display
</details>

</details>

</details>
Commit: 7ff54efdd2ad8f34db5d10efc4774ffe08ff9fe6

File: packages/next/src/export/routes/app-page.ts [hunk 1]
```diff
@@ -128,12 +128,23 @@ export async function exportAppPage(
       'utf8'
     )
 
+    const isParallelRoute = /\/@\w+/.test(page)
```
Comment (ijjk) [inline packages/next/src/export/routes/app-page.ts:131]: Are we prerendering parallel routes, if so why do we need special casing for them at this point? 

Also could we have a test case asserting this behavior possibly?

Comment (huozhi) [reply packages/next/src/export/routes/app-page.ts:131]: We need the flight data for prefetching or navigation. I’ll check if we have that later

Comment (huozhi) [reply packages/next/src/export/routes/app-page.ts:131]: add a test case for parallel routes static generation in [7826720](https://github.com/vercel/next.js/pull/63731/commits/782672008745cbba32d1291964d4dff33916b7cd)

**Review by ijjk — APPROVED** [review]

**Comment by avianion** [comment]
⨯ Failed to generate static paths for /chat/[id]:
log.js:80
Error: Jest worker encountered 2 child process exceptions, exceeding retry limit
    at ChildProcessWorker.initialize (/app/nextjs/node_modules/next/dist/compiled/jest-worker/index.js:1:11580)
    at ChildProcessWorker._onExit (/app/nextjs/node_modules/next/dist/compiled/jest-worker/index.js:1:12545)
    at ChildProcess.emit (C:\Users\dexim\Desktop\avian\lib\events.js:518:28)
    at ChildProcess.emit (C:\Users\dexim\Desktop\avian\lib\domain.js:488:12)
    at ChildProcess._handle.onexit (C:\Users\dexim\Desktop\avian\lib\internal\child_process.js:294:12)
    at Process.callbackTrampoline (node:internal/async_hooks:130:17) {type: 'WorkerError', stack: 'Error: Jest worker encountered 2 child proces…Trampoline (node:internal/async_hooks:130:17)', message: 'Jest worker encountered 2 child process exceptions, exceeding retry limit'}


@huozhi @hu0p @hallaji @ijjk @ztanner @feedthejim @wyattjoh 

This issue still occurs with dynamic page paths...heres the reproduction case #63740 
https://github.com/vercel/next.js/issues/63740

**Comment by nikcio** [comment]
Hi this didn't fully fix #62228 
See my comment here: https://github.com/vercel/next.js/issues/62228#issuecomment-2025084206

**Pull request merged**